src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -0,0 +1,188 @@
+
+package com.google.sps.servlets;
+import com.google.gson.*;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.QueryResultList;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.User;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp;
+import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
+import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
+import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.client.util.store.FileDataStoreFactory;
+
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.security.GeneralSecurityException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import com.google.sps.data.UserInfo;
+
+/** Servlet that returns events sorted by most recent timestamp */
+@WebServlet("/user")
+public class UserServlet extends HttpServlet {
+
+  static final String CHALLENGE_ID_PARAM = "chal";
+  static final String CHALLENGE_STATUS_PARAM = "stat";
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    
+    String userId = userService.getCurrentUser().getUserId();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query query = new Query(UserInfo.DATA_TYPE).setFilter(new FilterPredicate(UserInfo.ID, FilterOperator.EQUAL, userId));
+    Entity entity = datastore.prepare(query).asSingleEntity();
+
+    if (entity == null) {
+        doPost(request, response);
```

Comment:
This line is very strange.

1. doPost will trigger a response.sendRedirect. However, later in this function you do response.getWriter.println()
2. POST and GET have very different semantics. It is surprising to see a GET action alter the backend in any way.

To address (1) - I'd suggest you refactor the code which actually creates a new user somewhere, perhaps `UserInfo` or a private method of this class? Then you can call that refactored method in both `doGet` and `doPost` to avoid some of the response wrangling `doPost` does.

To address (2) - I'd suggest you make `doGet` return 404 when the user isn't created yet. Let the caller of `GET /user` handle what happens on the 404.